compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -449,6 +453,53 @@ class TreeChecker extends Phase with SymTransformer {
       tree
     }
   }
+
+  /**
+    * Checks that `New` nodes are always wrapped inside `Select` nodes.
+    */
+  private object WrappedNewChecker {
+
+    // Current state of the search. We return as soon as we see the first improperly wrapped `New`.
+    sealed trait State
+    case class NotFound(parent: Option[tpd.Tree]) extends State
+    case class Found(parent: Option[tpd.Tree], child: tpd.New) extends State
+
+    def assertSelectWrapsNew(tree: tpd.Tree)(implicit ctx: Context): Unit = {
+      val detect = new TreeAccumulator[State] {
+        override def apply(st: State, tree: tpd.Tree)(implicit ctx: Context): State = {
+          st match {
+            case _: Found => st // thread the found pair
+            case NotFound(oldParent) =>
+              tree match {
+                case tree: tpd.New if !canWrapNew(oldParent) => Found(oldParent, tree)
+                case _ =>
+                  // replace the parent when folding over the children
+                  foldOver(NotFound(Some(tree)), tree) match {
+                    case found: Found => found
+                    case _ => st // return the old parent so that my siblings see it
+                  }
+              }
+          }
+        }
+      }
+
+      detect(NotFound(None), tree) match {
+        case Found(Some(parent), child) =>
+          assert(assertion = false, i"`New` node must be wrapped in a `Select`:\n  parent = ${parent.show}\n  child = ${child.show}")
```

Comment:
If you do this you might not even need the parameter of `Found` and could make it a `case object`